core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java

```diff
@@ -910,6 +913,20 @@ public Optional<TableLayout> getNewTableLayout(Session session, String catalogNa
                 .map(layout -> new TableLayout(catalogHandle, transactionHandle, layout));
     }
 
+    @Override
+    public Optional<Type> coerceNewTableColumn(Session session, CatalogHandle catalogHandle, Type type)
+    {
+        CatalogMetadata catalogMetadata = getCatalogMetadata(session, catalogHandle);
+        ConnectorMetadata metadata = catalogMetadata.getMetadata(session);
+        return metadata.coerceNewTableColumn(session.toConnectorSession(catalogHandle), type)
+                .map(newType -> {
+                    if (!typeCoercion.isCompatible(newType, type)) {
+                        throw new TrinoException(NOT_SUPPORTED, format("Type '%s' is not compatible with the supplied type '%s' in coerceNewTableColumn", type, newType));
```

Comment:
This should be `FUNCTION_IMPLEMENTATION_ERROR`. It's a programming error for the connector to return a type that is not cast-compatible with the requested type.

Also, please add that requirement to the documentation of the SPI method.